### ISSUE TITLE
openssl FIPS_mode_set error

### ISSUE THREAD

**Issue by bimbimprasetyoafif** [spacemonkeygo/openssl#151]
i hope this repo not die yet. I face error when running my golang app that used this library.
it return 
`# github.com/spacemonkeygo/openssl
../../go/pkg/mod/github.com/spacemonkeygo/openssl@v0.0.0-20181017203307-c2dcc5cca94a/fips.go:31:7: could not determine kind of name for C.FIPS_mode_set
`
my thought is the version of openssl, this library used and mine was different.

note: running on ubuntu 22.04, openssl 3.0.2

**Comment by nisarg0103** [comment]
same issue @bimbimprasetyoafif , have you got any solution ?

**Comment by bimbimprasetyoafif** [comment]
> same issue @bimbimprasetyoafif , have you got any solution ?

Nope, i just downgrade to ubuntu 20 to solve my problem, I thought there's no dependency for latest version. Beside, this repo has been no update since 2018 @nisarg0103

**Comment by ashishm8898** [comment]
You need to downgrade your ubuntu version to 20 from 22...

**Comment by ashishm8898** [comment]
Now this comes in My macbook M1 chip

**Comment by ricky-charlet** [comment]
See my untested pull request for a possible solution.
https://github.com/spacemonkeygo/openssl/pull/154

**Comment by rodrigorodriguescosta** [comment]
same issue, is there any solution except downgrade ubuntu?

**Comment by huwcbjones** [comment]
FIPS_mode_set got removed in OpenSSL 3.
There is an OpenSSL 3 only fork that doesn't use deprecated symbols here: https://github.com/pexip/go-openssl

**Comment by Rocky210** [comment]
Is there any solution for this error 
# github.com/spacemonkeygo/openssl
../../../go/pkg/mod/github.com/spacemonkeygo/openssl@v0.0.0-20181017203307-c2dcc5cca94a/fips.go:31:7: could not determine kind of name for C.FIPS_mode_set

**Comment by huwcbjones** [comment]
@Rocky210 see my previous comment regarding deprecated symbols in OpenSSL3 here: https://github.com/spacemonkeygo/openssl/issues/151#issuecomment-1828163438

**Comment by Rocky210** [comment]
I need a clear demonstration for implementation ,could you please explain .

**Comment by huwcbjones** [comment]
> I need a clear demonstration for implementation ,could you please explain .

I will do my best to explain what's changed, point you in the right direction and provide you with some untested code snippets.

From the [OpenSSL 3 manpage](https://www.openssl.org/docs/man3.0/man7/fips_module.html), gone are `FIPS_mode()` and `FIPS_mode_set()`, hence this issue. They do not exist, you cannot call them.

OpenSSL 3 introduced an architecture change with the introduction of library contexts and providers.
If no provider is loaded and cryptographic functions are called, the "default" provider will be loaded.
Therefore users requiring programmatic enabling of the FIPS module should load the fips provider into the default library context before performing any crypto operations.

I've just had a look at what we've got in our OpenSSL 3 fork and it appears we load the default provider on init.
https://github.com/pexip/go-openssl/blob/master/init.go#L108
https://github.com/pexip/go-openssl/blob/60019a99ece1aea7302abbb6b9a6157252bac72a/provider.go#L26-L31

Because of how we use FIPS crypto in our product, we enable/disable FIPS mode on a VM level, so we do not do any programmatic loading. However, if I were to add support to our fork, I'd probably do something like in https://github.com/pexip/go-openssl/blob/master/provider.go

```go
func loadFIPSProvider() error {
	defaultCtx = &LibraryContext{
		ctx: nil, providers: make(map[string]*C.OSSL_PROVIDER), mu: &sync.Mutex{},
	}
	runtime.SetFinalizer(defaultCtx, func(c *LibraryContext) { c.finalise() })
	if err := defaultCtx.LoadProvider("fips"); err != nil {
		return fmt.Errorf("failed to load fips provider: %w", err)
	}
	if err := defaultCtx.LoadProvider("base"); err != nil {
		return fmt.Errorf("failed to load base provider: %w", err)
	}
	return nil
}
```
That snippet above roughly matches the spirit of the [C example in the manpage](https://www.openssl.org/docs/man3.0/man7/fips_module.html#Programmatically-loading-the-FIPS-module-default-library-context).

Then you'd have to fiddle with [`init.go/init`](https://github.com/pexip/go-openssl/blob/60019a99ece1aea7302abbb6b9a6157252bac72a/init.go#L104-L109) to either not load the default provider, thereby forcing the user to, or alternatively provide a function to unload the preloaded providers in the library context and re-initialise the default library context with the fips one.

Something like this should do the trick
```go
func LoadFIPSProvider() error {
	oldDefaultCtx := defaultCtx
	oldDefaultCtx.finalize()
	return loadFIPSProvider()
}
```
